csharp/src/Ice/ACM.cs

```diff
@@ -105,204 +162,125 @@ internal void Destroy()
                 }
 
                 _communicator = null;
                 _changes.Clear();
 
-                //
-                // Wait for the connection set to be cleared by the timer thread.
-                //
+                // Wait for the connection set to be cleared.
                 while (_connections.Count > 0)
                 {
-                    System.Threading.Monitor.Wait(_mutex);
+                    Monitor.Wait(_mutex);
                 }
             }
         }
 
-        public void Add(Connection connection)
+        internal IEnumerable<Connection> SwapReapedConnections()
         {
-            if (_config.Timeout == 0)
-            {
-                return;
-            }
-
             lock (_mutex)
             {
-                if (_connections.Count == 0)
-                {
-                    Debug.Assert(_communicator != null);
-                    _connections.Add(connection);
-                    _communicator.Timer().ScheduleRepeated(this, _config.Timeout / 2);
-                }
-                else
+                if (_reapedConnections.Count == 0)
                 {
-                    _changes.Add(new Change(connection, false));
+                    return Enumerable.Empty<Connection>();
                 }
+                List<Connection> connections = _reapedConnections;
+                _reapedConnections = new List<Connection>();
+                return connections;
             }
         }
 
-        public void Remove(Connection connection)
-        {
-            if (_config.Timeout == 0)
-            {
-                return;
-            }
-
-            lock (_mutex)
-            {
-                Debug.Assert(_communicator != null);
-                _changes.Add(new Change(connection, true));
-            }
-        }
-
-        public void Reap(Connection connection)
-        {
-            lock (_mutex)
-            {
-                _reapedConnections.Add(connection);
-            }
-        }
-
-        public IACMMonitor Acm(int? timeout, ACMClose? c, ACMHeartbeat? h)
-        {
-            Debug.Assert(_communicator != null);
-
-            ACMConfig config = _config.Clone();
-            if (timeout.HasValue)
-            {
-                config.Timeout = timeout.Value * 1000; // To milliseconds
-            }
-            if (c.HasValue)
-            {
-                config.Close = c.Value;
-            }
-            if (h.HasValue)
-            {
-                config.Heartbeat = h.Value;
-            }
-            return new ConnectionACMMonitor(this, _communicator.Timer(), config);
-        }
-
-        public ACM GetACM()
-        {
-            return new ACM
-            {
-                Timeout = _config.Timeout / 1000,
-                Close = _config.Close,
-                Heartbeat = _config.Heartbeat
-            };
-        }
-
-        internal IEnumerable<Connection> SwapReapedConnections()
+        internal void HandleException(Exception ex)
```

Comment:
I would get rid of this method and add the logging in the catch block.

Reply:
Fixed